wunderio/client-UK-FSA-beta
# FSA-915 - Removal of non-existing establishments

This change introduces a mechanism where a full set of establishments is fetched from API (`Establishments/basic` endpoint) and stored in:
 - as paged files in `public://api/[date]/fhrs_results_[page].json`
 - as FHRSID and date in `fsa_establishment_api_import` table.
 
On cron run (supposedly it runs every minute) the following steps are performed:
- Page number to fetch is determined from previous stored paged result .json file (if last json is `fhrs_results_003.json`, then API is queried with page 4) and results from API are downloaded to `public://api/[date]`
- Saved JSON file is parsed and FHRS values are merged into `fsa_establishment_api_import` table (FHRS values are unique, values are added or updated if they exist)
- When the full set of establishments are fetched:
  - entry `api_fetch_finish_last_date` is set to current date in `fsa_ratings_import` key/value collection (this allows skip the process of determining the next page, etc)
  - all values in `fsa_establishment_api_import` table are matched against migrated establishment entities in `migrate_map_fsa_establishment` tables and entities that are defined as migrated but not available in fetched set are removed as non-existing
  - entry `entity_purge_finish_last_date` is set to current date in `fsa_ratings_import` key/value collection (this allows skip the process of finding the result-set/migrated entity diff, etc)

Steps to review:
1. Run `drush cim -y` and `drush cr`
2. Run `drush updb` to create the `fsa_establishment_api_import` table
3. Check that cron is running every minute
4. In couple of minutes see that files are appearing in `public://api/[date]/`
5. When the all the files are fetched (usually count 106) see that `fsa_establishment_api_import` table has values
6. See that queue worker jobs named `fsa_ratings_import_entity_purger` are created in `queue` table
7. See that obsolete entities (entity IDs are included in queue job data) are removed
